python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,315 @@
+        if os.path.isfile(checkpoint.checkpoint_dir_or_data):
+            os.remove(checkpoint.checkpoint_dir_or_data)
+            return
+        elif os.path.isdir(checkpoint.checkpoint_dir_or_data):
+            shutil.rmtree(checkpoint.checkpoint_dir_or_data)
+            return
+    logger.warning(
+        f"Could not delete checkpoint {checkpoint} from disk as it is "
+        f"neither file not directory. Path: {checkpoint.checkpoint_dir_or_data}."
+    )
+
+
+class _HeapCheckpointWrapper:
+    def __init__(self, priority: Any, tracked_checkpoint: _TrackedCheckpoint):
+        self.priority = priority
+        self.tracked_checkpoint = tracked_checkpoint
+
+    def __lt__(self, other):
+        return self.priority < other.priority
+
+    def __repr__(self):
+        return f"_HeapCheckpoint({repr(self.tracked_checkpoint)})"
+
+
+@PublicAPI(stability="beta")
+@dataclass
+class CheckpointStrategy:
+    """Configurable parameters for defining the checkpointing strategy.
+
+    Default behavior is to persist all checkpoints to disk. If
+    ``num_to_keep`` is set, the default retention policy is to keep the
+    checkpoints with maximum timestamp, i.e. the most recent checkpoints.
+
+    Args:
+        num_to_keep (Optional[int]): The number of checkpoints to keep
+            on disk for this run. If a checkpoint is persisted to disk after
+            there are already this many checkpoints, then an existing
+            checkpoint will be deleted. If this is ``None`` then checkpoints
+            will not be deleted. If this is ``0`` then no checkpoints will be
+            persisted to disk.
+        checkpoint_score_attribute (str): The attribute that will be used to
+            score checkpoints to determine which checkpoints should be kept
+            on disk when there are greater than ``num_to_keep`` checkpoints.
+            This attribute must be a key from the checkpoint
+            dictionary which has a numerical value.
+        checkpoint_score_order (str). Either "max" or "min".
+            If "max", then checkpoints with highest values of
+            ``checkpoint_score_attribute`` will be kept.
+            If "min", then checkpoints with lowest values of
+            ``checkpoint_score_attribute`` will be kept.
+    """
+
+    num_to_keep: Optional[int] = None
+    checkpoint_score_attribute: str = TRAINING_ITERATION
+    checkpoint_score_order: str = MAX
+
+    def __post_init__(self):
+        if self.num_to_keep is not None and self.num_to_keep < 0:
+            raise ValueError(
+                f"Received invalidate num_to_keep: "
+                f"{self.num_to_keep}. "
+                f"Must be None or non-negative integer."
+            )
+        if self.checkpoint_score_order not in (MAX, MIN):
+            raise ValueError(
+                f"checkpoint_score_order must be either " f'"{MAX}" or "{MIN}".'
+            )
+
+
+class CheckpointManager:
```

Comment:
in the spirit of having more documentation in our code, can you write some description of how this Manager works? :)

Reply:
I had added a docstring in a recent commit, is that sufficient or do you think it requires more detail?